lib/cluster.rb

```diff
@@ -120,6 +120,34 @@ def format
     @format ||= CalculateFormat.new(self).cluster_format
   end
 
+  def organizations_in_cluster
+    @organizations_in_cluster ||= (holdings.pluck(:organization) +
+                                  ht_items.pluck(:billing_entity)).uniq
+  end
+
+  # Maps enumchrons to list of orgs that have an item with that enumchron
+  # Not necessary for SPMs/SERs
+  def item_enum_chron_orgs(enum)
```

Comment:
I'm worried that by caching the values (via `||=`) in all these organization-counting methods we'll be ignoring any additions to the cluster. Might not be a real issue if we're never doing both adds and analysis at the same time, but thought I'd bring it up as at least a candidate for documenting it.

```ruby
cluster.organizations_in_cluster
#=> ['anu', 'bc', 'cmu']
cluster.item_enum_chron_orgs(enum).include?('umich')
#=> false
cluster.add_ht_items(items_from_umich)
cluster.organizations_in_cluster
#=> still ['anu', 'bc', 'cmu']
cluster.item_enum_chron_orgs(enum).include?('umich')
#=> still always gonna be false
```

This potentially drags us down the garden path of dirty-record tracking, which no one wants. It could be mitigated by either zeroing out the cached values (like `@item_enum_chron_orgs`) in `add_ht_items`, or by just documenting and being careful. The former could be made easier if all these computed values were kept in a single data structure which could then be unassigned, making sure we don't miss any. Or maybe there's an `HtItemSet` class in here waiting to be defined, but that seems like it wouldn't mesh super well with all the Mongoid niceties around embedded docs.

I'd also set `*_enum_chron_orgs` to  `Hash.new{|k,h| h[k] = []}` in the initializer, instead of dragging that default-value-is-a-hash logic into each method, or maybe use a dead-simple hash-of-arrays class to make sure that you never get a scalar in there:

```ruby
class HashOfArrays < SimpleDelegator
  def initialize
    @h = Hash.new{|h,k| h[k] = []}
    self.__setobj__(@h)
  end
  def []=(k,v)
    @h[k] = [v]
  end
end

h = HashOfArrays.new
#=> {}
h[1] #=> []
h[2] << :two 
h[2] << :too 
h[3] = :three

 pp h
#=> {1=>[], 2=>[:two, :too], 3=>[:three]}`

```

Reply:
I *think* running cost report generation and record insertion at the same is mostly a non-starter for anything but guesstimation. Whether a cluster is altered in the midst of h_share calculation or 5 ms after is meaningless. If we are running a holdings update, we can't get any more precise than "cost report generated while in the middle of a holdings update" with either approach.